server/proxy/grpcproxy/watch.go

```diff
@@ -311,3 +316,27 @@ func (wps *watchProxyStream) delete(id int64) {
 	}
 	wps.watchCh <- resp
 }
+
+func (wps *watchProxyStream) requestProgressAll() error {
+	wps.mu.Lock()
+	defer wps.mu.Unlock()
+
+	for _, w := range wps.watchers {
+		// Enable the progress notify on the watcher, otherwise the send()
+		// method may drop progress events.
+		if !w.progress {
+			w.progress = true
```

Comment:
This is not the same progress notification source. In etcd there are two ways to get progress notification:
* Server configures a periodic progress notification period and client requests getting periodic notification. For example if it's configured for 5s, and there was no normal event in watch for 5s, server should send a notification to watch.
* Client requested notification, client can at any moment request a notification and server should imminently give a response about a watch progress.

Reply:
> Just to clarify, are you saying I shouldn't permanently enable progress notifications on the watcher?
Instead, should I implement something like a one-time toggle on the watcher side to allow a notification for the next request only? 

Yes

> So, when the proxy receives a new WatchRequest_ProgressRequest, it would set this toggle to true and call RequestProgress on the server.
Then, once the proxy receives the progress notification, it sends it to the client and resets the toggle back to false.
Is that the intended behavior?

I haven't dive into what exactly the toggle does, but my initial thinking is that grpcproxy should just proxy WatchRequest_ProgressRequest request.

Reply:
Thanks. 
I've reworked the code to support both types of progress notifications.
Let me know if this looks better.

Reply:
@serathius any thoughts on the new implementation ?